src/components/SideNavigation/SideNavigation.tsx

```diff
@@ -0,0 +1,183 @@
+import SideNavigationItem from "./SideNavigationItem";
+import type {
+  SideNavigationLinkDefaultElement,
+  SideNavigationLinkProps,
+} from "./SideNavigationLink";
+import { isReactNode } from "utils";
+
+export type NavItem<L = SideNavigationLinkDefaultElement> =
+  | SideNavigationItemProps<L>
+  | ReactNode
+  | null;
+
+export type NavItemGroup<L = SideNavigationLinkDefaultElement> = NavItem<L>[];
+
+export type NavGroup<L = SideNavigationLinkDefaultElement> = PropsWithSpread<
+  {
+    /**
+     * The navigation items.
+     */
+    items: NavItemGroup<L>;
+  },
+  HTMLProps<HTMLUListElement>
+>;
+
+export type Props<L = SideNavigationLinkDefaultElement> = PropsWithSpread<
+  {
+    /**
+     * The navigation content. This can be used instead of supplying `items`.
+     */
+    children?: PropsWithChildren["children"];
+    /**
+     * Whether to use the dark theme.
+     */
+    dark?: boolean;
+    /**
+     * Whether the navigation items contain icons.
+     */
+    hasIcons?: boolean;
+    /**
+     * The navigation items. This can be a single array of items, nested arrays
+     * for multiple navigation lists or an object containing items and props
+     * that are passed to the list element.
+     */
+    items?:
+      | NavItemGroup<L>
+      | (NavItemGroup<L> | null)[]
+      | (NavGroup<L> | null)[];
+    /**
+     * The component or element to use for the link elements e.g. `a` or `NavLink`.
+     * @default a
+     */
+    linkComponent?: SideNavigationLinkProps["component"];
+    /**
+     * Classes to apply to the navigation list(s).
+     */
+    listClassName?: string;
+    /**
+     * Classes to apply to the nav element.
+     */
+    navClassName?: string;
+  },
+  HTMLProps<HTMLDivElement>
+>;
+
+const generateItem = <L = SideNavigationLinkDefaultElement,>(
+  item: NavItem<L>,
+  index: number,
+  linkComponent?: Props<L>["linkComponent"],
+  dark?: boolean
+) => {
+  if (isReactNode(item)) {
+    return <SideNavigationItem key={index}>{item}</SideNavigationItem>;
+  }
+  if ("nonInteractive" in item) {
+    return (
+      <SideNavigationItem {...item} dark={item.dark ?? dark} key={index} />
+    );
+  }
+  if ("children" in item) {
+    return <SideNavigationItem key={index} {...item} />;
+  }
+  if ("label" in item) {
+    return (
+      <SideNavigationItem<L>
+        component={item.component ?? linkComponent}
+        dark={item.dark ?? dark}
+        {...item}
+        key={index}
+      />
+    );
+  }
+  return null;
+};
+
+const generateItems = <L = SideNavigationLinkDefaultElement,>(
+  groups: (NavItemGroup<L> | null)[] | (NavGroup<L> | null)[] | null,
+  listClassName?: string,
+  linkComponent?: SideNavigationLinkProps["component"],
+  dark?: boolean
+) => {
+  return groups
+    ?.filter(Boolean)
+    .map((group: NavItemGroup<L> | NavGroup<L>, g) => {
+      let items: NavItem<L>[];
+      let props: HTMLProps<HTMLUListElement> = {};
+      if (typeof group === "object" && "items" in group) {
```

Comment:
This is a bit confusing to me. Why can the group come in so many different shapes? Should we have a more strict approach here, so the group is of a specific well-defined type, instead of giving options to be out of various shapes?

Reply:
Yeah that's probably sensible. This type grew as I updated the side nav to support the different projects, but we could just use the object type which would support all cases e.g.
```typescript
    <SideNavigation
      items={[
        {
          className: "extra-ul-class",
          items: [
            {
              label: "Link one",
              href: "#",
            },
          ],
        },
        {
          items: [
            {
              label: "Link two",
              href: "#",
              icon: "user",
            },
          ],
        },
      ]}
    />
```

In this case, if you only had one ul you'd still have to provide the items in an object `items={[{ items: [...] }]}` but that's probably a reasonable trade-off to make the component generally easier to work with.

Does that sound better?

Reply:
it might be better to separate the SideNavigation and the SideNavigationItems, so SideNavigation only accepts children, which can be SideNavigationItems wrapped in one or multiple uls. That way we keep the SideNavigation component concise and don't have to come up with a generic interface for the items.

Reply:
The component supports `children` if you don't want to supply items, but you lose some benefits e.g. not being able to define the link component and dark/light mode in one place and have it passed to the nav items.

Reply:
I see, you can either pass children and take over or rely on a selection of props to have a more guided experience.The pattern seems to be newly introduced to react-components. When checking existing usage of `children` they seem to be the only option for components exposing them. 

I see the benefit of allowing flexibility with children and also allowing a shorter usage with the guided props. Though I am not a big fan of it, it makes the interface a bit unclear. When using children, some of the props will be ignored. When providing both children and guided props the behaviour becomes unclear as to what is actually going to be rendered. So maybe better to only have the flexible option with children? I am not opposing the idea completely, just my opinion at this poin. I would like to get other opinions on it.

Reply:
I've updated this to only allow items to be nested arrays of objects, or instead you pass children to replace the contents.